Test fail on LLVM15 `test_noinline_on_main_call`

```
% pytest numba/tests/test_codegen.py -k test_noinline_on_main_call -s
======================================================================================= test session starts =======================================================================================
platform darwin -- Python 3.11.8, pytest-7.4.4, pluggy-1.0.0
rootdir: /Users/siu/dev/numba
collecting ... Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
collected 9 items / 8 deselected / 1 selected

numba/tests/test_codegen.py ; ModuleID = 'TestWrappers.test_noinline_on_main_call.<locals>.foo'
source_filename = "<string>"
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-darwin23.3.0"

@".const.TestWrappers.test_noinline_on_main_call.<locals>.foo" = internal constant [53 x i8] c"TestWrappers.test_noinline_on_main_call.<locals>.foo\00"
@_ZN08NumbaEnv5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE = common local_unnamed_addr global i8* null
@".const.missing Environment: _ZN08NumbaEnv5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE" = internal constant [167 x i8] c"missing Environment: _ZN08NumbaEnv5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE\00"
@_Py_NoneStruct = external global i8
@PyExc_RuntimeError = external global i8

; Function Attrs: argmemonly mustprogress nofree norecurse nosync nounwind willreturn writeonly
define i32 @_ZN5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE(i8** noalias nocapture writeonly %retptr, { i8*, i32, i8*, i8*, i32 }** noalias nocapture readnone %excinfo) local_unnamed_addr #0 {
entry:
  store i8* null, i8** %retptr, align 8
  ret i32 0
}

define i8* @_ZN7cpython5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE(i8* nocapture readnone %py_closure, i8* %py_args, i8* nocapture readnone %py_kws) local_unnamed_addr {
entry:
  %.5 = tail call i32 (i8*, i8*, i64, i64, ...) @PyArg_UnpackTuple(i8* %py_args, i8* getelementptr inbounds ([53 x i8], [53 x i8]* @".const.TestWrappers.test_noinline_on_main_call.<locals>.foo", i64 0, i64 0), i64 0, i64 0)
  %.6 = icmp eq i32 %.5, 0
  br i1 %.6, label %common.ret, label %entry.endif, !prof !0

common.ret:                                       ; preds = %entry, %entry.endif.endif, %entry.endif.if
  %common.ret.op = phi i8* [ null, %entry.endif.if ], [ @_Py_NoneStruct, %entry.endif.endif ], [ null, %entry ]
  ret i8* %common.ret.op

entry.endif:                                      ; preds = %entry
  %.10 = load i8*, i8** @_ZN08NumbaEnv5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE, align 8
  %.15 = icmp eq i8* %.10, null
  br i1 %.15, label %entry.endif.if, label %entry.endif.endif, !prof !0

entry.endif.if:                                   ; preds = %entry.endif
  tail call void @PyErr_SetString(i8* nonnull @PyExc_RuntimeError, i8* getelementptr inbounds ([167 x i8], [167 x i8]* @".const.missing Environment: _ZN08NumbaEnv5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE", i64 0, i64 0))
  br label %common.ret

entry.endif.endif:                                ; preds = %entry.endif
  tail call void @Py_IncRef(i8* nonnull @_Py_NoneStruct)
  br label %common.ret
}

declare i32 @PyArg_UnpackTuple(i8*, i8*, i64, i64, ...) local_unnamed_addr

declare void @PyErr_SetString(i8*, i8*) local_unnamed_addr

declare void @Py_IncRef(i8*) local_unnamed_addr

; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn writeonly
define i8* @cfunc._ZN5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE() local_unnamed_addr #1 {
entry:
  %.2 = alloca i8*, align 8
  store i8* null, i8** %.2, align 8
  %.6 = call i32 @_ZN5numba5tests12test_codegen12TestWrappers26test_noinline_on_main_call12_3clocals_3e3fooB2v1B38c8tJTIeFIjxB2IKSgI4CrvQClQZ6FczSBAA_3dE(i8** nonnull %.2, { i8*, i32, i8*, i8*, i32 }** nonnull poison) #2
  %.16 = load i8*, i8** %.2, align 8
  ret i8* %.16
}

attributes #0 = { argmemonly mustprogress nofree norecurse nosync nounwind willreturn writeonly }
attributes #1 = { mustprogress nofree norecurse nosync nounwind willreturn writeonly }
attributes #2 = { noinline }

!0 = !{!"branch_weights", i32 1, i32 99}
F

============================================================================================ FAILURES =============================================================================================
_____________________________________________________________________________ TestWrappers.test_noinline_on_main_call _____________________________________________________________________________

self = <numba.tests.test_codegen.TestWrappers testMethod=test_noinline_on_main_call>

    def test_noinline_on_main_call(self):
        # Checks that the cpython and cfunc wrapper produces a call with the
        # "noinline" attr present for the decorated function.

        @njit
        def foo():
            pass

        foo()
        sig = foo.signatures[0]

        ol = foo.overloads[sig]
        name = ol.fndesc.mangled_name.replace("$", r"\$")
        p1 = r".*call.*{}".format(name)
        p2 = r".*(#[0-9]+).*"
        call_site = re.compile(p1 + p2)

        lines = foo.inspect_llvm(sig).splitlines()
        print('\n'.join(lines))
        meta_data_idx = []
        for l in lines:
            matched = call_site.match(l)
            if matched:
                meta_data_idx.append(matched.groups()[0])

        # should be 2 calls, one from cpython wrapper one from cfunc wrapper
>       self.assertEqual(len(meta_data_idx), 2)
E       AssertionError: 1 != 2

numba/tests/test_codegen.py:238: AssertionError
===================================================================================== short test summary info =====================================================================================
FAILED numba/tests/test_codegen.py::TestWrappers::test_noinline_on_main_call - AssertionError: 1 != 2
================================================================================= 1 failed, 8 deselected in 1.15s =================================================================================
```

The test checks that the function is not inlined into either wrappers. Strangely, the `noinline` attribute is on the function as it can be seen in OPT=0, but with OPT>0 the cpython wrapper version is inlining the call anyway. cfunc wrapper didn't though.

--- a/numba/tests/npyufunc/test_errors.py
+++ b/numba/tests/npyufunc/test_errors.py
@@ -6,7 +6,7 @@ import numpy as np
 from numba import vectorize, guvectorize
 
 from numba.tests.support import (TestCase, CheckWarningsMixin,
-                                 skip_m1_fenv_errors)
+                                 skip_macos_fenv_errors)
 import unittest
 
 
@@ -116,7 +116,7 @@ class TestFloatingPointExceptions(TestCase, CheckWarningsMixin):
                                 ["divide by zero encountered",
                                  "invalid value encountered"])
 
-    @skip_m1_fenv_errors
+    @skip_macos_fenv_errors
     def test_remainder_float(self):
         self.check_divmod_float(remainder,
                                 [0.0, float('nan'), float('nan'), 1.0],

--- a/numba/tests/support.py
+++ b/numba/tests/support.py
@@ -176,9 +176,9 @@ skip_ppc64le_invalid_ctr_loop = unittest.skipIf(
 # fenv.h on M1 may have various issues:
 # https://github.com/numba/numba/issues/7822#issuecomment-1065356758
 _uname = platform.uname()
-IS_OSX_ARM64 = _uname.system == 'Darwin' and _uname.machine == 'arm64'
-skip_m1_fenv_errors = unittest.skipIf(IS_OSX_ARM64,
-    "fenv.h-like functionality unreliable on OSX arm64")
+IS_MACOS = _uname.system == 'Darwin'
+skip_macos_fenv_errors = unittest.skipIf(IS_MACOS,
+    "fenv.h-like functionality unreliable on macOS")
 
 try:
     import scipy.linalg.cython_lapack

--- a/numba/tests/support.py
+++ b/numba/tests/support.py
@@ -179,6 +179,7 @@ _uname = platform.uname()
 IS_MACOS = _uname.system == 'Darwin'
 skip_macos_fenv_errors = unittest.skipIf(IS_MACOS,
     "fenv.h-like functionality unreliable on macOS")
+IS_MACOS_ARM64 = IS_MACOS and _uname.machine == 'arm64'
 
 try:
     import scipy.linalg.cython_lapack

--- a/numba/tests/test_codegen.py
+++ b/numba/tests/test_codegen.py
@@ -213,9 +213,12 @@ class TestWrappers(TestCase):
         # Checks that the cpython and cfunc wrapper produces a call with the
         # "noinline" attr present for the decorated function.
 
+        # Creating a refcounted object induces a side-effect that prevents IPO
+        # from eliding the call to the function with LLVM 15. See Issue #9658:
+        # https://github.com/numba/numba/issues/9658
         @njit
         def foo():
-            pass
+            return list([1])
 
         foo()
         sig = foo.signatures[0]

--- a/numba/tests/test_np_functions.py
+++ b/numba/tests/test_np_functions.py
@@ -21,7 +21,7 @@ from numba.np.extensions import cross2d
 from numba.tests.support import (TestCase, MemoryLeakMixin,
                                  needs_blas, run_in_subprocess,
                                  skip_if_numpy_2, IS_NUMPY_2,
-                                 IS_OSX_ARM64)
+                                 IS_MACOS_ARM64)
 import unittest
 
 
@@ -4664,7 +4664,7 @@ class TestNPFunctions(MemoryLeakMixin, TestCase):
             got = cfunc(x, xp, fp)
             self.assertPreciseEqual(expected, got, abs_tol=atol)
 
-    @unittest.skipIf(IS_NUMPY_2 and IS_OSX_ARM64, "NEP 50 interaction issue.")
+    @unittest.skipIf(IS_NUMPY_2 and IS_MACOS_ARM64, "NEP 50 interaction issue.")
     def test_interp_complex_stress_tests(self):
         pyfunc = interp
         cfunc = jit(nopython=True)(pyfunc)

